atmcfg/atmcfg.go

```diff
@@ -44,37 +56,111 @@ func setDisabledByReplicaSetName(out *opsmngr.AutomationConfig, name string, dis
 		for _, m := range rs.Members {
 			for k, p := range out.Processes {
 				if p.Name == m.Host {
-					out.Processes[k].Disabled = disabled
+					setDisable(out.Processes[k], processesMap, disabled)
 				}
 			}
 		}
 	}
 }
 
-func setDisabledByShardName(out *opsmngr.AutomationConfig, name string, disabled bool) {
+func setDisabledByShardNameAndProcesses(out *opsmngr.AutomationConfig, name string, processesMap map[string]bool, disabled bool) {
 	i, found := search.ShardingConfig(out.Sharding, func(s *opsmngr.ShardingConfig) bool {
 		return s.Name == name
 	})
 	if found {
 		s := out.Sharding[i]
 		// disable shards
 		for _, rs := range s.Shards {
-			setDisabledByReplicaSetName(out, rs.ID, disabled)
+			setDisabledByReplicaSetNameAndProcesses(out, rs.ID, processesMap, disabled)
 		}
 		// disable config rs
-		setDisabledByReplicaSetName(out, s.ConfigServerReplica, disabled)
+		setDisabledByReplicaSetNameAndProcesses(out, s.ConfigServerReplica, processesMap, disabled)
 		// disable mongos
 		for i := range out.Processes {
 			if out.Processes[i].Cluster == name {
-				out.Processes[i].Disabled = disabled
+				setDisable(out.Processes[i], processesMap, disabled)
 			}
 		}
 	}
 }
 
+func setDisable(process *opsmngr.Process, processesMap map[string]bool, disabled bool) {
+	if len(processesMap) > 0 {
+		key := fmt.Sprintf("%s:%d", process.Hostname, process.Args26.NET.Port)
+		if _, ok := processesMap[key]; ok {
```

Comment:
not blocking but you could do `if _, ok := processesMap[key]; ok || len(processesMap) == 0 {` to eliminate one `if` here

Reply:
I do not see how to remove the if. I have to check that `len(processesMap)>0` before using `processesMap[key]`

Reply:
how is right there, `if _, ok := processesMap[key]; ok || len(processesMap) == 0` if `processesMap` is empty then ok is false but  len(processesMap) == 0 is true, `false` or `true` is `true` you just need to make sure `processesMap` is not nil or you'll get a pointer error